src/audio_buffer.rs

```diff
@@ -0,0 +1,482 @@
+                for sample in reader.samples::<i16>() {
+                    let s = sample.unwrap() as f32 / i16::MAX as f32;
+                    assert!(s.abs() < 1.);
+                    decoded[channel_number].push(s);
+                    // next sample belongs to the next channel
+                    channel_number = (channel_number + 1) % number_of_channels;
+                }
+            } else {
+                panic!("bits_per_sample {:?} not implemented", bits_per_sample);
+            }
+        }
+        hound::SampleFormat::Float => {
+            // this one is not tested
+            let mut channel_number = 0;
+
+            for sample in reader.samples::<f32>() {
+                let s = sample.unwrap();
+                decoded[channel_number].push(s);
+                // next sample belongs to the next channel
+                channel_number = (channel_number + 1) % number_of_channels;
+            }
+        }
+    }
+
+    assert_eq!(decoded[0].len(), length); // duration is ok
+                                          // println!("decoded length: {} - input length: {}", decoded[0].len(), length);
+
+    // [spec] Take the result, representing the decoded linear PCM audio data,
+    // and resample it to the sample-rate of the BaseAudioContext if it is
+    // different from the sample-rate of audioData.
+
+    // @todo - resample if needed
+    // if (sample_rate != self.sample_rate()) {
+    //     // @todo - resample
+    //     // cf. https://github.com/HEnquist/rubato/blob/master/examples/fftfixedin64.rs
+    //     // already used in waveshaper
+    // }
+
+    AudioBuffer {
+        number_of_channels,
+        length,
+        // @TODO - this is wrong, should be the AudioContext.sampleRate
+        sample_rate: SampleRate(sample_rate),
+        internal_data: AudioBufferData::from(decoded),
+    }
+}
+
+// `AudioBufferData` is basically a simple matrix representing an aligned,
+// de-interleaved "in memory" audio asset (i.e. a complete sound file), i.e.:
+// 0 (left) : [0.0, 1.0, ...]
+// 1 (right): [0.1, 0.9, ...]
+// ...
+//
+// @note: each channel is an `Arc` that can be cloned by the audio nodes, we
+// only need Arc and not Mutex here because if a node acquired a channel,
+// it must keep a reference to it even if `set_channel_data` is called. In such
+// situation we just replace the channel with a new `Arc` pointing to new `Vec`.
+// So no mutation occurs and memory allocation takes place in the control thread.
+// Having the `Arc` at the level of the single channel, will probably lead to a
+// bit more bookkeeping in the node side, but allows to be a bit more efficent in
+// terms of memory, as we can discard only one channel at a time.
+//
+// @note - spec does not define "audio asset" which can lead to confusion
+//
+// @see - <https://webaudio.github.io/web-audio-api/#acquire-the-content>
+#[derive(Clone)]
+pub(crate) struct AudioBufferData {
+    channels: Rc<Vec<Arc<Vec<f32>>>>,
```

Comment:
Mixing Rc en Arc is really bad, as the result is neither thread safe nor efficient.
Just ask yourself, does this data need to cross thread boundaries (e.g. from control to render)? Then use Arc

Reply:
I let you decide on that point